### DISCUSSION TITLE
Launch multiple async commands or spawn multiple threads from one command

### DISCUSSION THREAD

**Discussion by azerpas** [tauri-apps/tauri#3459] [Q&A] [answered]
Hi,

I'm looking to execute a lot (up to 1000) of simultaneous similar tasks that will last a few seconds.

I've thought about:
- Calling multiple Tauri async commands from the React side with Promise.all
- Calling one Tauri async command that will spawn the needed threads 

What would be the benefits and disadvantages of each one?

**Comment by FabianLars** [comment, marked as answer]
Generally i'd say the "call one command and do it all in rust" is better for multi-threaded workloads, especially since js being single threaded and the communication channel most likely being a bottleneck here.
Also, multi-threading is one of rust main selling points, might as well use it :P You have more control and (in most cases) better performance this way.

The main exception here is if you would be bottlenecked by the communication channel anyway (as in: you return huge amounts of data from the commands), then it's probably more like personal preference (good idea to benchmark that i guess).

tldr: huge amount of tasks and you don't need to return much data? -> Rust

**Reply by azerpas** [reply]
Thank you for the detailed answer. :)

One could also use Tauri Events to "return" data to the front on each thread end, right?

**Reply by FabianLars** [reply]
Yeah, absolutely

**Reply by evbo** [reply]
@FabianLars Does choice of async runtime affect this answer? For instance, spawning threads that originate from a single Tauri Command, I presume that means using something like `tokio`'s `join_all` expression to concurrently wait for a set of futures to resolve.

But unless I'm mistaken, that would come at the cost of an additional dependency on `tokio` (significant footprint). Whereas if JS code called dozens of individual async tauri commands I wouldn't need something like `tokio`

Am I totally getting this wrong, or is there some built-in async runtime Tauri comes with off the shelf and if you decide to do your own spawing of threads that comes at the cost of a larger bundle size?

**Reply by evbo** [reply]
Also, I read the docs on Tauri Events, what exactly would the pros/cons be in relation to this question?:
>  It is analogous to the command system, but payload type check must be written on the event handler and it simplifies communication from the backend to the frontend, working like a channel.

To put it crudely, is it _faster_ than Commands for this sort of use case discussed by OP? I dug around the discord but only saw 1 post from you saying window events are generally handled with Tauri events, so a bit of clarification appreciated :)

**Reply by FabianLars** [reply]
@evbo 
> Does choice of async runtime affect this answer?

Hmm iirc i had mainly actual threads in mind when i answered, but no, not really, especially since you don't have much of a choice since tauri already includes tokio (accessible via `tauri::async_runtime` btw).

> is it faster than Commands for this sort of use case discussed by OP?

There shouldn't be much of a difference (not noticeable, maybe not even realistically measurable), because both use the same IPC system.
